### PR TITLE
Fix inferred union types for Ds Map::get() and Map::remove()

### PR DESCRIPTION
**Pull request by simPod**
👋 can you guide me with this please?

I'm passing `Map<int, int|string>` into the function.

However, return type of `get(0)` call on the map is resolved as `int`. I'd expect `int|string`.

```php
/** @param Map<int, int|string> $a */
public function mapGetUnionType(Map $a) : void
{
	assertType('int|string', $a->get(0));
}
```

```
Expected :'int|string'
Actual   :'int'
```

Map stub is 

```php
/**
 * @template TKey
 * @template TValue
 * @implements Collection<TKey, TValue>
 */
final class Map implements Collection
{
    /**
     * @param iterable<TKey, TValue>|null $values
     */
    public function __construct(?iterable $values = null)
    {
    }
```

So I believe I'm saying that `TValue` is `int|string`.

`Map::get()`'s signature is 

```php
/**
 * @template TDefault
 * @param TKey $key
 * @param TDefault $default
 * @return TValue|TDefault
 * @throws OutOfBoundsException
 */
public function get($key, $default = null)
```

so `get(0)` should return `TValue` (`int|string`). Or not? Thanks!

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
No idea what's wrong, What happens if you pass in `$a->get('foo')`?

**Comment by ondrejmirtes** [comment]
Oh, I realized what's wrong. There's this extension: https://github.com/phpstan/phpstan-src/blob/master/src/Type/Php/DsMapDynamicReturnTypeExtension.php

You can try to remove it and see what happens.

**Comment by simPod** [comment]
You're right. Would it make sense to drop it then. Like in the commit 👆

File: tests/PHPStan/Analyser/data/ext-ds.php [hunk 1]
```diff
@@ -16,6 +16,12 @@ class B
 
 class Foo
 {
+	/** @param Map<int, int|string> $a */
+	public function mapGetUnionType(Map $a) : void
+	{
+		assertType('int|string|TDefault (method Ds\Map::get(), parameter)', $a->get(1));
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/ext-ds.php:22]: This isn't ideal. The result should be `int|string|null`. We should resolve TDefault to the default parameter value.

That's logic that should be done here https://github.com/phpstan/phpstan-src/blob/9b712a90e3fa03a0d26b51ec2ff35c807ab06d11/src/Reflection/GenericParametersAcceptorResolver.php#L33 but isn't. Can you look into it? My theory is that ext-ds reflection gives us wrong data.

Comment (simPod) [reply tests/PHPStan/Analyser/data/ext-ds.php:22]: We should achieve 
`assertType('int|string|', $a->get(1));`  
`assertType('int|string|null', $a->get(1, null));`

Can't do that with just template annotations I believe. I'll try to edit the extension.

Comment (ondrejmirtes) [reply tests/PHPStan/Analyser/data/ext-ds.php:22]: What happens when `$a->get(1)` is not found?

**Comment by simPod** [comment]
OutOfBounds exception is thrown if I remember correctly

**Comment by ondrejmirtes** [comment]
You're right. So the `@return TValue|TDefault` is wrong if the second argument is not present. So the extension must be modified and not deleted.

File: src/Type/Php/DsMapDynamicReturnTypeExtension.php [hunk 2]
```diff
@@ -35,11 +35,11 @@ public function getTypeFromMethodCall(MethodReflection $methodReflection, Method
 			)->getReturnType();
 		}
 
-		if (! $returnType instanceof UnionType) {
-			return $returnType;
+		if ($returnType instanceof UnionType) {
+			return new UnionType(array_slice($returnType->getTypes(), 0, -1));
```
Comment (ondrejmirtes) [inline src/Type/Php/DsMapDynamicReturnTypeExtension.php:39]: This doesn't seem right at all.

1) You can't assume what will be the last element in an array in a UnionType.
2) UnionType of a single element is invalid.

Instead, you should get `->getTypes()`, filter the unwanted type (I guess `NullType`?) and then call `TypeCombinator::union(...$types)` on the rest.

Comment (simPod) [reply src/Type/Php/DsMapDynamicReturnTypeExtension.php]: With the new commit I filter out TDefault (I assume if present it means no value was provided)

**Comment by ondrejmirtes** [comment]
Looks fine, thank you!

**Pull request merged**